File: school-backend/src/main/java/com/school/sms/util/ValidationPatterns.java
package com.school.sms.util;

/**
 * Regex constants shared by the request DTOs.
 *
 * They live in one place because the alternative was nine independently-written
 * copies of the same rule, which had already drifted: most DTOs demanded
 * {@code ^\+?[0-9]{10,15}$} — digits only — while every phone number the
 * application actually stores looks like {@code +91-9810011122}. The result was
 * that a record could be read from the database and then rejected when saved
 * back unchanged.
 */
public final class ValidationPatterns {

    private ValidationPatterns() {
    }

    /**
     * Phone numbers as they are really stored and entered: digits, optionally with
     * {@code + - ( )} and spaces.
     *
     * Deliberately permissive. Phone formatting varies by country and by data
     * source, and a validator stricter than the data it guards does not improve
     * quality — it just makes existing records uneditable. The empty alternative
     * lets optional fields be cleared; required fields pair this with
     * {@code @NotBlank}.
     *
     * Kept identical to PHONE_PATTERN in
     * {@code school-frontend/src/utils/validationPatterns.ts}, so the client and
     * the API agree on what they will accept.
     */
    public static final String PHONE = "^$|^[0-9+\\-\\s()]{6,20}$";

    public static final String PHONE_MESSAGE =
            "Phone number may contain digits and optionally + - ( ) and spaces (6-20 characters)";

    /** Indian PIN code — six digits, matching what the seed data and forms use. */
    public static final String PINCODE = "^$|^[0-9]{4,10}$";

    public static final String PINCODE_MESSAGE = "Pincode must be 4-10 digits";
}

File: school-frontend/src/utils/validationPatterns.ts
/**
 * Validation patterns shared by the forms that edit existing records.
 *
 * These exist because a stricter rule than the data actually stored makes a
 * record uneditable: the form loads a value it considers invalid, validation
 * blocks submit on a field the user never touched, and the save button appears
 * to do nothing.
 */

/**
 * Phone numbers as they are actually stored — `+91-9810011122` throughout the
 * seed data, as well as plain `9810011122`.
 *
 * A digits-only `^[0-9]{10}$` rule rejects every seeded record. It also
 * contradicts the backend, whose own rule (StudentSelfUpdateRequest,
 * TeacherSelfUpdateRequest) is this same permissive shape — so the stricter
 * client rule rejected values the API would have accepted.
 *
 * Public-facing forms that only ever take fresh input (registration, admission
 * and contact enquiries) keep the stricter 10-digit rule on purpose: there is no
 * stored value to contradict, and a tighter rule is better at the point of entry.
 */
export const PHONE_PATTERN = /^[0-9+\-\s()]{6,20}$/;
export const PHONE_MESSAGE = 'Enter a valid phone number (digits, and optionally + - spaces)';

/** Indian PIN code. Seed data uses six digits, so this one matches reality already. */
export const PINCODE_PATTERN = /^[0-9]{6}$/;
export const PINCODE_MESSAGE = 'Enter a valid 6-digit pincode';
